File: frontend/src/api/planCulture.ts
import client from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface PlanVariete {
  id_plan_variete: number
  id_packgraine: number
  nb_plantes: number
  taille_pot_l?: number
  ordre: number
  // Enrichis
  nom_variete?: string
  nom_breeder?: string
  type_graine?: string
  duree_flo_min?: number
  duree_flo_max?: number
  stock_disponible: number
  paquet_ouvert: boolean
  prix_achat_pack?: number
  date_achat_pack?: string
  duree_conservation_mois?: number
}

export interface PlanCulture {
  id_plan: number
  nom: string
  id_espace?: number
  nom_espace?: string
  surface_m2?: number
  statut: 'brouillon' | 'pret' | 'lance'
  notes?: string
  created_at?: string
  updated_at?: string
  varietes: PlanVariete[]
  nb_plantes_total: number
}

export interface PlanCultureCreate {
  nom: string
  id_espace?: number
  notes?: string
}

export interface PlanCultureUpdate {
  nom?: string
  id_espace?: number
  statut?: string
  notes?: string
}

export interface PlanVarieteCreate {
  id_packgraine: number
  nb_plantes?: number
  taille_pot_l?: number
}

export interface PlanVarieteUpdate {
  nb_plantes?: number
  taille_pot_l?: number
}

export interface CatalogueItem {
  id_packgraine: number
  nom_variete?: string
  nom_breeder?: string
  type_graine?: string
  duree_flo_min?: number
  duree_flo_max?: number
  stock_disponible: number
  paquet_ouvert: boolean
  prix_achat?: number
  date_achat?: string
  duree_conservation_mois?: number
  nbr_graines_total?: number
}

// ─── API calls ────────────────────────────────────────────────────────────────

export const planCultureAPI = {
  // Plans
  getAll: () => client.get<PlanCulture[]>('/plan-culture/'),
  getOne: (id: number) => client.get<PlanCulture>(`/plan-culture/${id}`),
  create: (data: PlanCultureCreate) => client.post<PlanCulture>('/plan-culture/', data),
  update: (id: number, data: PlanCultureUpdate) => client.put<PlanCulture>(`/plan-culture/${id}`, data),
  delete: (id: number) => client.delete(`/plan-culture/${id}`),

  // Variétés d'un plan
  addVariete: (planId: number, data: PlanVarieteCreate) =>
    client.post<PlanVariete>(`/plan-culture/${planId}/varietes`, data),
  updateVariete: (planId: number, pvId: number, data: PlanVarieteUpdate) =>
    client.put<PlanVariete>(`/plan-culture/${planId}/varietes/${pvId}`, data),
  removeVariete: (planId: number, pvId: number) =>
    client.delete(`/plan-culture/${planId}/varietes/${pvId}`),

  // Export CSV
  exportCSV: (planId: number) => {
    // Déclenche le téléchargement directement (pas d'axios — on veut un vrai download)
    const baseUrl = (client.defaults.baseURL || '').replace(/\/$/, '')
    const url = `${baseUrl}/plan-culture/${planId}/export/csv`
    const a = document.createElement('a')
    a.href = url
    a.download = ''
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
  },

  // Calcul pots
  calcNbPots: (surface_m2: number, taille_pot_l: number) =>
    client.get<{ nb_pots_recommande: number }>('/plan-culture/utils/nb-pots', {
      params: { surface_m2, taille_pot_l },
    }),

  // Catalogue filtrable
  getCatalogue: (params?: {
    breeder?: string
    variete?: string
    type_graine?: string
    flo_min?: number
    flo_max?: number
    stock_seulement?: boolean
  }) => client.get<CatalogueItem[]>('/plan-culture/utils/catalogue', { params }),
}

File: backend/app/schemas/plan_culture.py
"""Schemas Pydantic pour PlanCulture et PlanCultureVariete"""
from pydantic import BaseModel, ConfigDict
from typing import Optional
from datetime import datetime


# ─── Variété dans un plan ─────────────────────────────────────────────────────

class PlanVarieteCreate(BaseModel):
    id_packgraine: int
    nb_plantes: int = 1
    taille_pot_l: Optional[float] = None
    ordre: int = 0


class PlanVarieteUpdate(BaseModel):
    nb_plantes: Optional[int] = None
    taille_pot_l: Optional[float] = None
    ordre: Optional[int] = None


class PlanVarieteRead(BaseModel):
    id_plan_variete: int
    id_packgraine: int
    nb_plantes: int
    taille_pot_l: Optional[float] = None
    ordre: int = 0
    # Champs enrichis depuis Graine / PackGraine
    nom_variete: Optional[str] = None
    nom_breeder: Optional[str] = None
    type_graine: Optional[str] = None        # reguliere | auto | fem
    duree_flo_min: Optional[int] = None
    duree_flo_max: Optional[int] = None
    stock_disponible: int = 0                # nb graines restantes dans le pack
    paquet_ouvert: bool = False              # au moins une graine utilisée
    prix_achat_pack: Optional[float] = None
    date_achat_pack: Optional[str] = None
    duree_conservation_mois: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


# ─── Plan de culture ──────────────────────────────────────────────────────────

class PlanCultureCreate(BaseModel):
    nom: str
    id_espace: Optional[int] = None
    notes: Optional[str] = None


class PlanCultureUpdate(BaseModel):
    nom: Optional[str] = None
    id_espace: Optional[int] = None
    statut: Optional[str] = None
    notes: Optional[str] = None


class PlanCultureRead(BaseModel):
    id_plan: int
    nom: str
    id_espace: Optional[int] = None
    nom_espace: Optional[str] = None
    surface_m2: Optional[float] = None
    statut: str = "brouillon"
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    varietes: list[PlanVarieteRead] = []
    nb_plantes_total: int = 0   # somme nb_plantes de toutes les lignes

    model_config = ConfigDict(from_attributes=True)


# ─── Calcul pots ──────────────────────────────────────────────────────────────

class NbPotsResult(BaseModel):
    surface_m2: float
    taille_pot_l: float
    nb_pots_recommande: int
